docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -1,184 +1,231 @@
-Kokkos::parallel_for(... {
-  new((Derived*)deviceInstance) Derived(); // construct an instance in the place, the pointer deviceInstance points to
-});
+class Base {
+  public:
+  void Foo() {}
+
+  KOKKOS_FUNCTION
+  virtual void Bar() {}
+};
+
+class Derived : public Base {
+  public:
+  KOKKOS_FUNCTION
+  void Bar() override {}
+};
+
+int main(int argc, char *argv[])
+{
+  Kokkos::initialize(argc, argv);
+  {
+
+  // create
+  void* deviceInstanceMemory = Kokkos::kokkos_malloc(sizeof(Derived)); // allocate memory on device
+  Kokkos::parallel_for("initialize", 1, KOKKOS_LAMBDA (const int i) {
+    new (static_cast<Derived*>(deviceInstanceMemory)) Derived(); // initialize on device
+  });
+  Base* deviceInstance = static_cast<Derived*>(deviceInstanceMemory); // declare on this memory
+
+  // use
+  Kokkos::parallel_for("myKernel", 10, KOKKOS_LAMBDA (const int i) {
+      deviceInstance->Bar();
+  });
+
+  // cleanup
+  Kokkos::parallel_for("destroy", 1, KOKKOS_LAMBDA (const int i) {
+    deviceInstance->~Base(); // destroy on device
+  });
+  Kokkos::fence("wait for destroy");
+  Kokkos::kokkos_free(deviceInstanceMemory); // free
+
+  }
+  Kokkos::finalize();
+}
 ```
 
+We first use the `KOKKOS_FUNCTION` macro to make the methods callable from a kernel.
+When creating the instance, note that we introduce a distinction between the *memory* that the it uses, and the actual instantiated *object*.
+The construct is done on the device, within a single-iteration `parallel_for`, using placement new.
+Since the kernel does not have a return type, we use a static cast to associate the object with the memory allocation.
+
+The destructor must explicitly be called from the device, again with a single-iteration `parallel_for`. Then the memory allocation can be release with `kokkos_free`.
+
 This code is extremely ugly, but leads to functional virtual function calls on the device. The Vpointer now points to the device Vtable.
+Remember that those virtual functions cannot be called on the host anymore!
 
 ![VPointerToDevice](./figures/VirtualFunctions-VPointerToDevice.png)
 
-Note that like with other uses of `new`, you need to later `free` the memory.
 For a full working example, see [the example in the repo](https://github.com/kokkos/kokkos/blob/master/example/virtual_functions/main.cpp).
 
-## Complications and Fixes
+## What if I need a setter that works with host values?
 
-The first problem people run into with this is that they want to initialize some fields or nested classes based on host data before moving data down to the device
+The first problem people run into with this is when they want to initialize some fields based on host data, using a setter which is *not* a virtual function.
+Calling this setter on the device would crash if the host data doesn't form a valid type on the device: especially, if the data type is not copyable on the device (e.g. for a pointer), or if the data instance is invalid on the device.
```

Comment:
```suggestion
The first problem people run into with this is when they want to set some fields based on host data. As the object instance resides in device memory, it might not be accessible by the host. But the fields can be set within a `parallel_for` on the device. Nevertheless, this requires that the lambda or functor that sets the fields on the device must have access to the host data.
```